[Presto-on-Spark Native] Join query fails with Multiple entries with same key: orderkey=9 and orderkey=0

A join query that returns duplicate column names fails:

```
SELECT * FROM orders o, lineitem l WHERE o.orderkey = l.orderkey;

Caused by: java.lang.IllegalArgumentException: Multiple entries with same key: orderkey=9 and orderkey=0
	at com.google.common.collect.ImmutableMap.conflictException(ImmutableMap.java:215)
	at com.google.common.collect.ImmutableMap.checkNoConflict(ImmutableMap.java:209)
	at com.google.common.collect.RegularImmutableMap.checkNoConflictInKeyBucket(RegularImmutableMap.java:147)
	at com.google.common.collect.RegularImmutableMap.fromEntryArray(RegularImmutableMap.java:110)
	at com.google.common.collect.ImmutableMap$Builder.build(ImmutableMap.java:393)
	at com.facebook.presto.sql.planner.LocalExecutionPlanner$CustomPlanTranslator.makeLayoutFromOutputVariables(LocalExecutionPlanner.java:886)
	at com.facebook.presto.sql.planner.LocalExecutionPlanner$CustomPlanTranslator.makeLayout(LocalExecutionPlanner.java:875)
	at com.facebook.presto.spark.execution.operator.NativeExecutionOperator$NativeExecutionOperatorTranslator.translate(NativeExecutionOperator.java:408)
	at com.facebook.presto.sql.planner.LocalExecutionPlanner$Visitor.visitPlan(LocalExecutionPlanner.java:3123)
	at com.facebook.presto.sql.planner.LocalExecutionPlanner$Visitor.visitPlan(LocalExecutionPlanner.java:900)
```

The "duplicate" column is "orderkey" present in both lineitem and orders.

CC: @rschlussel @miaoever @shrinidhijoshi

--- a/presto-main/src/main/java/com/facebook/presto/sql/planner/BasePlanFragmenter.java
+++ b/presto-main/src/main/java/com/facebook/presto/sql/planner/BasePlanFragmenter.java
@@ -181,6 +181,29 @@ public abstract class BasePlanFragmenter
 
         // Only delegate non-coordinatorOnly plan fragment to native engine
         if (isNativeExecutionEnabled(session) && !properties.getPartitioningHandle().isCoordinatorOnly()) {
+            if (root instanceof OutputNode) {
+                // OutputNode is special in that it can have duplicate output variables since it
+                // does not get converted to a PhysicalOperation for execution.
+                // Regular plan nodes, including NativeExecutionNode, must have unique output variables.
+                // Check if OutputNode has duplicate output variables and remove these.
+                // This is safe because OutputNode variables are not used by the workers.
+                OutputNode outputNode = (OutputNode) root;
+                List<VariableReferenceExpression> outputVariables = outputNode.getOutputVariables();
+                List<VariableReferenceExpression> newOutputVariables = new ArrayList<>();
+                List<String> newColumnNames = new ArrayList<>();
+                Set<VariableReferenceExpression> uniqueOutputVariables = new HashSet<>();
+                for (int i = 0; i < outputVariables.size(); ++i) {
+                    VariableReferenceExpression variable = outputVariables.get(i);
+                    if (uniqueOutputVariables.add(variable)) {
+                        newOutputVariables.add(variable);
+                        newColumnNames.add(outputNode.getColumnNames().get(i));
+                    }
+                }
+
+                if (uniqueOutputVariables.size() < outputVariables.size()) {
+                    root = new OutputNode(outputNode.getSourceLocation(), outputNode.getId(), outputNode.getSource(), newColumnNames, newOutputVariables);
+                }
+            }
             root = new NativeExecutionNode(root);
             schedulingOrder = scheduleOrder(root);
         }

--- a/presto-native-execution/src/test/java/com/facebook/presto/spark/TestPrestoSparkNativeExecution.java
+++ b/presto-native-execution/src/test/java/com/facebook/presto/spark/TestPrestoSparkNativeExecution.java
@@ -60,7 +60,7 @@ public class TestPrestoSparkNativeExecution
     @Test
     public void testJoins()
     {
-        assertQuery("SELECT count(*) FROM orders o, lineitem l WHERE o.orderkey = l.orderkey AND o.orderkey % 2 = 1");
+        assertQuery("SELECT * FROM orders o, lineitem l WHERE o.orderkey = l.orderkey AND o.orderkey % 2 = 1");
     }
 
     @Test

